jetty-client/src/main/java/org/eclipse/jetty/client/util/SPNEGOAuthentication.java

```diff
@@ -0,0 +1,319 @@
+        this.renewTGT = renewTGT;
+    }
+
+    @Override
+    public Result authenticate(Request request, ContentResponse response, HeaderInfo headerInfo, Attributes context)
+    {
+        SPNEGOContext spnegoContext = (SPNEGOContext)context.getAttribute(SPNEGOContext.ATTRIBUTE);
+        if (LOG.isDebugEnabled())
+            LOG.debug("Authenticate with context {}", spnegoContext);
+        if (spnegoContext == null)
+        {
+            spnegoContext = login();
+            context.setAttribute(SPNEGOContext.ATTRIBUTE, spnegoContext);
+        }
+
+        String b64Input = headerInfo.getBase64();
+        byte[] input = b64Input == null ? new byte[0] : Base64.getDecoder().decode(b64Input);
+        byte[] output = Subject.doAs(spnegoContext.subject, initGSSContext(spnegoContext, request.getHost(), input));
+        String b64Output = output == null ? null : new String(Base64.getEncoder().encode(output));
+
+        // The result cannot be used for subsequent requests,
+        // so it always has a null URI to avoid being cached.
+        return new SPNEGOResult(null, b64Output);
+    }
+
+    private SPNEGOContext login()
+    {
+        try
+        {
+            // First login via JAAS using the Kerberos AS_REQ call, with a client user.
+            // This will populate the Subject with the client user principal and the TGT.
+            String user = getUserName();
+            if (LOG.isDebugEnabled())
+                LOG.debug("Logging in user {}", user);
+            CallbackHandler callbackHandler = new PasswordCallbackHandler();
+            LoginContext loginContext = new LoginContext("", null, callbackHandler, new SPNEGOConfiguration());
+            loginContext.login();
+            Subject subject = loginContext.getSubject();
+
+            SPNEGOContext spnegoContext = new SPNEGOContext();
+            spnegoContext.subject = subject;
+            if (LOG.isDebugEnabled())
+                LOG.debug("Initialized {}", spnegoContext);
+            return spnegoContext;
+        }
+        catch (LoginException x)
+        {
+            throw new RuntimeException(x);
+        }
+    }
+
+    private PrivilegedAction<byte[]> initGSSContext(SPNEGOContext spnegoContext, String host, byte[] bytes)
+    {
+        return () ->
+        {
+            try
+            {
+                // The call to initSecContext with the service name will
+                // trigger the Kerberos TGS_REQ call, asking for the SGT,
+                // which will be added to the Subject credentials because
+                // initSecContext() is called from within Subject.doAs().
+                GSSContext gssContext = spnegoContext.gssContext;
+                if (gssContext == null)
+                {
+                    String principal = getServiceName() + "@" + host;
+                    GSSName serviceName = gssManager.createName(principal, GSSName.NT_HOSTBASED_SERVICE);
+                    Oid spnegoOid = new Oid("1.3.6.1.5.5.2");
```

Comment:
What is this magic number about?

Reply:
https://docs.oracle.com/javase/8/docs/technotes/guides/security/jgss/lab/part5.html


In order to use SPNEGO, you only need to specify the SPNEGO OID as follows:

``
Oid spnegoOid = new Oid("1.3.6.1.5.5.2");
``

Reply:
Ah. TIL.